clang/lib/AST/PreorderAST.cpp

```diff
@@ -278,229 +233,133 @@ void PreorderAST::AddZero(Expr *E, Node *Parent) {
   Create(E, /*Parent*/ N);
 }
 
-void PreorderAST::Coalesce(Node *N, bool &Changed) {
+bool BinaryOperatorNode::CanCoalesce() {
+  // We can only coalesce if the operator of the current and parent node is
+  // commutative and associative. This is because after coalescing we later
+  // need to sort the nodes and if the operator is not commutative and
+  // associative then sorting would be incorrect.
+  if (!IsOpCommutativeAndAssociative())
+    return false;
+  auto *BParent = dyn_cast_or_null<BinaryOperatorNode>(Parent);
+  if (!BParent || !BParent->IsOpCommutativeAndAssociative())
+    return false;
+
+  // We can coalesce in the following scenarios:
+  // 1. The current and parent nodes have the same operator OR
+  // 2. The current node is the only child of its operator node (maybe as a
+  // result of constant folding).
+  return Opc == BParent->Opc || Children.size() == 1;
+}
+
+void BinaryOperatorNode::Coalesce(bool &Changed, bool &Error) {
   if (Error)
     return;
 
-  if (!N)
-    return;
+  // Coalesce the children first.
+  // Since Children is modified within the loop, we need to evaluate
+  // the loop end on each iteration.
+  for (size_t I = 0; I != Children.size(); ++I) {
+    auto *Child = Children[I];
+    Child->Coalesce(Changed, Error);
+  }
 
-  // TODO: GitHub checkedc-clang issue #1032. Each kind of Node should have
-  // its own Coalesce method.
-  switch (N->Kind) {
-    default:
-      break;
-    case Node::NodeKind::BinaryOperatorNode: {
-      auto *B = dyn_cast<BinaryOperatorNode>(N);
+  if (!CanCoalesce())
+    return;
 
-      // Coalesce the children first.
-      for (auto *Child : B->Children)
-        Coalesce(Child, Changed);
+  // If the current node can be coalesced, its parent must be a
+  // BinaryOperatorNode.
+  auto *BParent = dyn_cast_or_null<BinaryOperatorNode>(Parent);
+  if (!BParent)
+    return;
 
-      if (CanCoalesceNode(B)) {
-        CoalesceNode(B);
-        Changed = true;
-      }
-      break;
-    }
-    case Node::NodeKind::UnaryOperatorNode: {
-      auto *U = dyn_cast<UnaryOperatorNode>(N);
-      Coalesce(U->Child, Changed);
-      break;
-    }
-    case Node::NodeKind::MemberNode: {
-      auto *M = dyn_cast<MemberNode>(N);
-      Coalesce(M->Base, Changed);
-      break;
-    }
-    case Node::NodeKind::ImplicitCastNode: {
-      auto *I = dyn_cast<ImplicitCastNode>(N);
-      Coalesce(I->Child, Changed);
+  // Remove the current node from the list of children of its parent.
+  // Since BParent is modified within the loop, we need to evaluate
+  // the loop end on each iteration.
+  for (auto I = BParent->Children.begin(); I != BParent->Children.end(); ++I) {
```

Comment:
Actually, as soon as the loop is modified we break out of the loop. So we don't really need to evaluate `BParent->Children.end()` on each iteration. This can be changed to:
`for (auto I = BParent->Children.begin(), E != BParent->Children.end(); I != E; ++I)`